rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,345 @@
+*not* be trusted according to the algorithm specified in [updating],
+we'd get an error like the following if we run `nix flake update`:
+```
+* **Failed trust verification!** 'someflake': 'git+https://github.com/some/flake.git?ref=master&rev=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' -!> 'git+https://github.com/some/flake.git?ref=master&rev=bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb' (Not signed by a trusted party!)
+```
+
+Notably, there is nothing that tells the user how to bypass the check
+using the `insecureUpdate` attribute,
+since if there were, users might very often ignore the error even if valid.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Merging PRs can not be done through GitHub.
+- Setting up PGP signatures will be a bit extra work for the committers.
+- Committers will need to maintain the list of authorized committers.
+- Fetching Nixpkgs with Nix will take around [5 times](#speeding-up-fetching-git) the time, depending on your hardware.
+- Currently building a fork of Nixpkgs (for e.g. a PR) will still work fine, since
+  the latest commit will be trusted when adding a new flake input, or referring
+  to one from the commandline, but updating the flake input will fail
+  since the commits are likely not signed by authorized committers.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Mandating signing only tags is an alternative, though not very useful for most users, since:
+  1) Users of Nixpkgs rarely fix their used revision to a tag.
+  2) Not signing every commit prevents you from using the algorithm described in [updating].
+  Somewhat the same thing can be achieved by the "redirection flake" design described
+  in [future], however.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should `/.authorized-committers.nix` perhaps not support computation like `flake.nix`?
+  If it supports computation, in Nixpkgs it could be derived from the maintainer list partially.
+- Should we require that repositories that use this mechanism contain
+  the public keys corresponding to the fingerprints to remove the need for key servers?
+
+# Future work
+[future]: #future-work
+
+- Currently, Git still uses SHA-1, which means that you can't
+  completely trust signatures on SHA-1 repositories, since it is a combination
+  of the commit message, parent hashes, and tree hashes, that are seemingly signed.
+  To avoid this, all Git repositories must eventually be transitioned to
+  SHA-256 when Git support for SHA-256 is stabilized. Users can already
+  try SHA-256 if they want extra security for their own repositories.
+  See:
+  - https://git-scm.com/docs/hash-function-transition/
+  - [Git 2.29.0 release notes](https://github.com/git/git/blob/master/Documentation/RelNotes/2.29.0.txt)
+  - [SHA-1 is a Shambles](https://eprint.iacr.org/2020/014.pdf)
+  - https://sha-mbles.github.io/ for a more readable version.
+- The `github` flake input type could potentially be removed since it doesn't seem to be necessary anymore.
```

Comment:
They are still required for fetching any repository that doesn't to do the whole signing game.

Reply:
That isn't true, they're just faster. The point of removing the github fetcher is that it would be unnecessary complexity.

Reply:
Yeah and for backwards compatibility they are also still required. We introduced them already (in an unversioned flake format....) and taking them away now would break every reproducibility guarantee Nix offered to those poor souls.

Reply:
Ah, no. `github:$X/$Y?Z` would just be a shorthand for `git+https://github.com/X/Y?$Z`. I'll make this clear.